scipy/stats/_hypotests.py

```diff
@@ -1262,30 +1262,46 @@ def _get_binomial_log_p_value_with_nuisance_param(
     return -log_pvalue
 
 
-def _pval_cvm_2samp_exact(s, nx, ny):
+def _pval_cvm_2samp_exact(s, m, n):
     """
     Compute the exact p-value of the Cramer-von Mises two-sample test
-    for a given value s (float) of the test statistic by enumerating
-    all possible combinations. nx and ny are the sizes of the samples.
+    for a given value s of the test statistic.
+    m and n are the sizes of the samples.
+
+    [1] Y. Xiao, A. Gordon, and A. Yakovlev, "A C++ Program for
+        the Cramér-Von Mises Two-Sample Test", J. Stat. Soft.,
+        vol. 17, no. 8, pp. 1-15, Dec. 2006.
+    [2] T. W. Anderson "On the Distribution of the Two-Sample Cramer-von Mises
+        Criterion," The Annals of Mathematical Statistics, Ann. Math. Statist.
+        33(3), 1148-1159, (September, 1962)
     """
-    rangex = np.arange(nx)
-    rangey = np.arange(ny)
 
-    us = []
-
-    # x and y are all possible partitions of ranks from 0 to nx + ny - 1
-    # into two sets of length nx and ny
-    # Here, ranks are from 0 to nx + ny - 1 instead of 1 to nx + ny, but
-    # this does not change the value of the statistic.
-    for x, y in _all_partitions(nx, ny):
-        # compute the statistic
-        u = nx * np.sum((x - rangex)**2)
-        u += ny * np.sum((y - rangey)**2)
-        us.append(u)
-
-    # compute the values of u and the frequencies
-    u, cnt = np.unique(us, return_counts=True)
-    return np.sum(cnt[u >= s]) / np.sum(cnt)
+    # [1, p. 3]
+    lcm = np.lcm(m, n)
+    # [1, p. 4], below eq. 3
+    a = lcm // m
+    b = lcm // n
+    # eq. 9 in [2] and eq. 2 in [1]
+    mn = m * n
+    zeta = lcm ** 2 * (m + n) * (6 * s - mn * (4 * mn - 1)) // (6 * mn ** 2)
+    # the frequency table of $g_{u, v}^+$ defined in [1, p. 6]
+    gs = [np.array([[0], [1]])] + [np.empty((2, 0), int) for _ in range(m)]
+    for u in range(n + 1):
+        next_gs = []
+        tmp = np.empty((2, 0), int)
+        for v, g in enumerate(gs):
+            # calculate g recursively with eq. 11 in [1]
```

Comment:
Would it be easy to just use recursion with a function then? I did that in `mannwhitneyu`; [here](https://github.com/scipy/scipy/blob/1f2a6e8549966983ab35fd4096f883dfe668a969/scipy/stats/_mannwhitneyu.py#L19) is the source for that implementing an algorithm from [this](https://www.sciencedirect.com/science/article/pii/S0378375898002614) article. Might be worth taking a look.

Reply:
In the first iteration of the outer loop (`u = 0`), $g_{0, 0}^+, g_{0, 1}^+, \ldots, g_{0, m}^+$ are calculated and stored in `gs`. We use the relation: $$\sum_{i = 1}^v (ai)^2 = (a \cdot 1)^2 + (a\cdot 2)^2 + \cdots + (av)^2 = a^2 v(v + 1)(2v + 1)/6.$$ Using the word in the article, $g_{0, v}^+ = S_{(av)^2}(g_{0, v - 1}^+)$ holds for $v > 0$: we can obtain $g_{0, v}^+$ by shifting $g_{0, v - 1}^+$ by $(av)^2$. This is how the code works:

1. `tmp` is at first empty (L. 1291).
2. In the first iteration of the inner loop (`v = 0`), `tmp` is merged with `gs[0]` (= `[[0], [1]]`), and then `(a * v - b * u) ** 2` (= `0`) is added to `tmp[0]`. Therefore `tmp` $= g_{0, 0}^+$ here.
3. In the remaining iterations of the inner loop, `tmp` is merged with `gs[v]`, which is empty, and then `(a * v - b * u) ** 2` (= `(a * v) ** 2`) is added to `tmp[0]`. If the `tmp` obtained in the previous loop was equal to $g_{0, v - 1}^+$, we get `tmp` $= g_{0, v}^+$ here.

Similary, $g_{u, 0}^+ = S_{(bu)^2}(g_{u - 1, 0}^+)$ holds. The first iteration of the outer loop (`u = 0`) gives us `gs[0]` $= g_{0, 0}^+$ (shown above). In the first iteration (`v = 0`) in the remaining iterations (`u > 0`), `tmp` is merged with `gs[0]` and `(a * v - b * u) ** 2` (= `(b * u) ** 2`) is added to `tmp[0]`. Therefore we have `gs[0]` $= g_{u, 0}^+$.